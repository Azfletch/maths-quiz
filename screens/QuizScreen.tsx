import React, { useState, useEffect } from 'react'
import { SafeAreaView, Text, View , TextInput, Pressable } from 'react-native'
import { Formik } from 'formik'

import questions from '../data/questions'
import TenFrame from '../components/TenFrame'

import { Question, ResultsAnswer, ScreenProps } from '../types'

const QuizScreen = ({ navigation }: ScreenProps): JSX.Element => {
  // Initialize State
  const [answer, setAnswer] = useState<number | null>(null)
  const [points, setPoints] = useState<number>(0)
  const [index, setIndex] = useState<number>(0)
  const [answerStatus, setAnswerStatus] = useState<boolean | null>(null)
  const [answers, setAnswers] = useState<ResultsAnswer[]>([])
  const [counter, setCounter] = useState<number>(15)

  let interval: any = null

 // Handle correct answer 
  useEffect(() => {
    if (Number(answer) === Number(currentQuestion?.correctAnswer)) {
      setPoints((points => points + 1))
      setAnswerStatus(true)
      answers.push({ question: index + 1, answer: true })
      
    } else {
      setAnswerStatus(false)
      answers.push({ question: index + 1, answer: false })
    }
  }, [answer])

  // Use answer status to reset state for next question
  useEffect(() => {
    if (answerStatus !== null) {
      setIndex(index + 1)
      setCounter(15)
      setAnswer(null)
    }
  }, [answerStatus])

  // Reset answer status
  useEffect(() => {
    setAnswerStatus(null)
  }, [index])

  // Handle timer
  useEffect(() => {
    const myInterval = () => {
      if (counter >= 1) {
        setCounter((state) => state - 1)
      }

      if (counter === 0) {
        setIndex(index + 1)
        setCounter(15)
      }
    }

    interval = setTimeout(myInterval, 1000)

    // clean up
    return () => {
      clearTimeout(interval)
    }
  }, [counter])

  // Reset timer
  useEffect(() => {
    if (!interval) {
      setCounter(15)
    }
  }, [index])

  // Handle navigation after last question
  useEffect(() => {
    const { length } = questions
    if(index + 1 > length) {

      clearTimeout(interval)
      navigation.navigate('Results', {
        answers: answers,
        points: points
      })
    }
  }, [index])

  // Reset timer
  useEffect(() => {
    if (!interval) {
      setCounter(15)
    }
  }, [index])

  const currentQuestion: Question = questions[index]

  return (
    <SafeAreaView>
      <View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between', padding: 10 }}>
        <Text testID='quiz-title' style={{ color: '#d257d2', fontWeight: '600', fontSize: 20 }}>MATHS CHALLENGE</Text>
        <Text testID='quiz-timer' style={{ padding: 10, backgroundColor: '#d057d2', borderRadius: 10, color: 'white', fontWeight: '600', fontSize: 20 }}>{counter}</Text>
      </View>

      <View style={{ flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between', marginHorizontal: 10 }}>
        <Text style={{ color: '#d257d2', fontWeight: '600', fontSize: 15 }}>Your Progress</Text>
        <Text style={{ color: '#d257d2', fontWeight: '600', fontSize: 15 }}>({index}/{questions.length}) questions answered</Text>
      </View>

      <View style={{ backgroundColor: '#f2f6a2', marginTop: 20, marginHorizontal: 10, padding: 15, borderRadius: 10 }}>
        <Text style={{ textAlign: 'center', color: '#d257d2', fontWeight: '600', fontSize: 20, marginBottom: 20 }}>{currentQuestion?.question}</Text>
        <TenFrame question={currentQuestion} />
        <View style={{ marginTop: 20 }}>
          <Formik
            initialValues={{ answer: '' }}
            onSubmit={async (values) => setAnswer(Number(values.answer))}>
            {({
              handleChange,
              handleBlur,
              handleSubmit,
              values,
              errors,
              touched,
            }) => (
              <View>
                <TextInput
                  style={{ backgroundColor: 'white', padding: 15, borderRadius: 10 }}
                  placeholder="Answer"
                  onChangeText={handleChange('answer')}
                  onBlur={handleBlur('answer')}
                  value={values.answer}
                />
                <Pressable
                  style={{ backgroundColor: '#d057d2', padding: 15, borderRadius: 15, marginLeft: 'auto', marginRight: 'auto', marginTop: 15 }}
                  onPress={handleSubmit}
                >
                  <Text style={{ color: 'white', fontWeight: '600', fontSize: 20 }}>SUBMIT</Text>
                </Pressable>
              </View>
            )}
          </Formik>
        </View>
      </View>
    </SafeAreaView>
  )
}

export default QuizScreen