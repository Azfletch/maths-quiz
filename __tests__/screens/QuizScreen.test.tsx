
import React from 'react'
import { render } from '@testing-library/react-native'

import QuizScreen from '../../screens/QuizScreen'

describe('QuizScreen', () => {
  it('renders correctly', () => {
    // Arrange
    const navigate = jest.fn()

    // Act
    const quizScreen = render(<QuizScreen navigation={{ navigate }} />)
    const timer = quizScreen.getByTestId('quiz-timer')
    const title = quizScreen.getByTestId('quiz-title')

    // Assert
    expect(quizScreen).toMatchSnapshot()
    expect(timer).toBeDefined()
    expect(title).toBeDefined()
  })
})
