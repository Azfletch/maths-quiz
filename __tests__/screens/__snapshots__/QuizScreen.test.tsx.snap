// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizScreen renders correctly 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#d257d2",
          "fontSize": 20,
          "fontWeight": "600",
        }
      }
      testID="quiz-title"
    >
      MATHS CHALLENGE
    </Text>
    <Text
      style={
        {
          "backgroundColor": "#d057d2",
          "borderRadius": 10,
          "color": "white",
          "fontSize": 20,
          "fontWeight": "600",
          "padding": 10,
        }
      }
      testID="quiz-timer"
    >
      15
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginHorizontal": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#d257d2",
          "fontSize": 15,
          "fontWeight": "600",
        }
      }
    >
      Your Progress
    </Text>
    <Text
      style={
        {
          "color": "#d257d2",
          "fontSize": 15,
          "fontWeight": "600",
        }
      }
    >
      (
      0
      /
      5
      ) questions answered
    </Text>
  </View>
  <View
    style={
      {
        "backgroundColor": "#f2f6a2",
        "borderRadius": 10,
        "marginHorizontal": 10,
        "marginTop": 20,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#d257d2",
          "fontSize": 20,
          "fontWeight": "600",
          "marginBottom": 20,
          "textAlign": "center",
        }
      }
    >
      How many blue counters?
    </Text>
    <View>
      <View
        style={
          {
            "flexDirection": "row",
            "marginLeft": "auto",
            "marginRight": "auto",
          }
        }
      >
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "red",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "red",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "red",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "marginLeft": "auto",
            "marginRight": "auto",
          }
        }
      >
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "blue",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "blue",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "blue",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "black",
              "borderWidth": 2,
              "padding": 20,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "blue",
                "borderRadius": 100,
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginTop": 20,
        }
      }
    >
      <View>
        <TextInput
          onBlur={[Function]}
          onChangeText={[Function]}
          placeholder="Answer"
          style={
            {
              "backgroundColor": "white",
              "borderRadius": 10,
              "padding": 15,
            }
          }
          value=""
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#d057d2",
              "borderRadius": 15,
              "marginLeft": "auto",
              "marginRight": "auto",
              "marginTop": 15,
              "padding": 15,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 20,
                "fontWeight": "600",
              }
            }
          >
            SUBMIT
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
